### PR TITLE
Feature 13448

### PR DESCRIPTION
**Pull request by mmoqui**
* The content of the comments isn't anymore used to set the index field `TITLE` and, through it, the index field `HEADER` dedicated to the search.
* A new facet is introduced to select only the comments among the search results.

Merge also the PR https://github.com/Silverpeas/Silverpeas-Components/pull/843

### CODE REVIEW

**Comment by mmoqui** [comment]
Rebased and merged by hand

**Comment by mmoqui** [comment]
Wrong PR closed!

**Review by SilverYoCha — CHANGES_REQUESTED** [review]

File: core-library/src/main/java/org/silverpeas/core/index/indexing/model/IndexEntryKey.java [hunk 1]
```diff
@@ -51,89 +61,121 @@ public final class IndexEntryKey implements Serializable {
   private static final String SEP = "|";
 
   /**
-   * The three parts of an IndexEntryKey are private and fixed at construction time.
+   * The four parts of an IndexEntryKey are private and fixed at construction time.
    */
   private final String component;
   private final String objectType;
   private final String objectId;
+  private final String linkedObjectId;
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/index/indexing/model/IndexEntryKey.java:69]: In fact, in my mind, a linked object is not a container object. A group or a space are container objects. A publication is a linked object as other contributions can be linked to it.
An IndexEntryKey refers always a given Silverpeas resource into an index. So, the component identifier, the type of the resource and the unique identifier of such a resource makes a reference to this resource in Silverpeas. Such reference attributes are enough to identify uniquely a first-class resource in Silverpeas (space, app, publication, ...) but for second-class resources such as comments or attachments. Latter require to have also the unique identifier of the object to which they are linked.

File: core-library/src/main/java/org/silverpeas/core/index/indexing/model/IndexEntryKey.java [hunk 2]
```diff
@@ -157,4 +202,9 @@ public boolean equals(Object o) {
   public int hashCode() {
     return toString().hashCode();
   }
+
+  private String normalize(final String raw) {
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/index/indexing/model/IndexEntryKey.java:206]: Yes, you're right. I'm going to change that.

File: core-services/comment/src/main/java/org/silverpeas/core/comment/service/DefaultCommentService.java [hunk 3]
```diff
@@ -197,23 +198,23 @@ public void unindexAllCommentsOnPublication(final String resourceType,
   }
 
   private void createIndex(final Comment cmt) {
-    int titleLength = 30;
+    User user = cmt.getCreator();
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-services/comment/src/main/java/org/silverpeas/core/comment/service/DefaultCommentService.java:201]: I'll check.

File: core-war/src/main/java/org/silverpeas/web/pdc/control/PdcSearchSessionController.java [hunk 4]
```diff
@@ -997,11 +955,12 @@ private void setExtraInfoToResultsToDisplay(List<GlobalSilverResult> results) {
           int iStart = underLink.indexOf(VERSIONING_RESOURCE);
           int iEnd = underLink.indexOf('&', iStart);
           underLink = underLink.substring(0, iStart) + PUBLICATION_RESOURCE +
-              underLink.substring(iEnd, underLink.length());
+              underLink.substring(iEnd);
           titleLink = buildTitleLink(markAsReadJS, downloadLink, underLink, true);
         } else if ("LinkedFile".equals(resultType)) {
           // open the linked file inside a popup window
-          downloadLink = FileServerUtils.getUrl(result.getName(), result.getId(), componentId);
+          String id = result.isLinked() ? result.getLinkedResourceId() : result.getId();
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-war/src/main/java/org/silverpeas/web/pdc/control/PdcSearchSessionController.java:962]: The code here is to satisfy the previous behavior: the 
```java
result.getId()
```
returned the identifier of the linked object (the publication or whatever). Indeed, previously the identifier of the indexed contribution (here the attachment) were set in the object type.

So the code here is to check the result refers a second-class contribution (the attachment) and, in this case, in order to have the same behavior than before, it is necessary to get the identifier of the linked object (id est publication or whatever)

File: core-war/src/main/java/org/silverpeas/web/pdc/control/PdcSearchSessionController.java [hunk 5]
```diff
@@ -1391,13 +1332,13 @@ public Jargon getJargon() {
...
     final String synonymsQueryString;
-    if (queryString == null || queryString.equals("") || !isThesaurusEnableByUser) {
+    if (queryString == null || queryString.isEmpty() || !isThesaurusEnableByUser) {
       synonymsQueryString = queryString;
     } else {
       StreamTokenizer st = new StreamTokenizer(new StringReader(queryString));
       st.resetSyntax();
       st.lowerCaseMode(true);
-      st.wordChars('\u0000', '\u00FF');
+      st.wordChars('\u0000', 'ÿ');
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-war/src/main/java/org/silverpeas/web/pdc/control/PdcSearchSessionController.java:1341]: Yes with Java 11

File: core-services/comment/src/main/java/org/silverpeas/core/comment/service/DefaultCommentService.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-services/comment/src/main/java/org/silverpeas/core/comment/service/DefaultCommentService.java:201]: I checked.

The content matching the search is returned, whatever the language in which it is written. In the case of the comments, their title in the results are in the language of the author of the comment. So, a French user behind a search will see the comments authored by an English user in the result page with their title in English.

In the other side, if the comments are written in English and the user asks for a text written in French, he shouldn't see the English comments unless there are similar words in both language.

[the remark this answers is not in the document]
Comment (SilverYoCha) [reply core-services/comment/src/main/java/org/silverpeas/core/comment/service/DefaultCommentService.java:201]: OK. I'll also be looking at that when I do my tests.

**Pull request merged**